src/GPUActivityAgent.cpp

```diff
@@ -0,0 +1,460 @@
+        }
+
+        // Policy provided initial values
+        double f_max = in_policy[M_POLICY_GPU_FREQ_MAX];
+        double f_efficient = in_policy[M_POLICY_GPU_FREQ_EFFICIENT];
+        double phi = in_policy[M_POLICY_GPU_PHI];
+
+        // initial range is needed to apply phi
+        double f_range = f_max - f_efficient;
+
+        if (phi > 0.5) {
+            //Energy Biased.  Scale F_max down to F_efficient based upon phi value
+            //Active region phi usage
+            f_max = std::max(f_efficient, f_max - f_range * (phi-0.5) / 0.5);
+        }
+        else if (phi < 0.5) {
+            //Perf Biased.  Scale F_efficient up to F_max based upon phi value
+            //Active region phi usage
+            f_efficient = std::min(f_max, f_efficient + f_range * (0.5-phi) / 0.5);
+        }
+
+        //Update Policy
+        in_policy[M_POLICY_GPU_FREQ_MAX] = f_max;
+        in_policy[M_POLICY_GPU_FREQ_EFFICIENT] = f_efficient;
+
+    }
+
+    // Distribute incoming policy to children
+    void GPUActivityAgent::split_policy(const std::vector<double>& in_policy,
+                                    std::vector<std::vector<double> >& out_policy)
+    {
+        assert(in_policy.size() == M_NUM_POLICY);
+        for (auto &child_pol : out_policy) {
+            child_pol = in_policy;
+        }
+    }
+
+    // Indicate whether to send the policy down to children
+    bool GPUActivityAgent::do_send_policy(void) const
+    {
+        return true;
+    }
+
+    void GPUActivityAgent::aggregate_sample(const std::vector<std::vector<double> > &in_sample,
+                                        std::vector<double>& out_sample)
+    {
+
+    }
+
+    // Indicate whether to send samples up to the parent
+    bool GPUActivityAgent::do_send_sample(void) const
+    {
+        return false;
+    }
+
+    void GPUActivityAgent::adjust_platform(const std::vector<double>& in_policy)
+    {
+        assert(in_policy.size() == M_NUM_POLICY);
+
+        m_do_write_batch = false;
+
+        // Per GPU freq
+        std::vector<double> gpu_freq_request;
+
+        // Values after phi has been applied
+        m_f_max = in_policy[M_POLICY_GPU_FREQ_MAX];
+        m_f_efficient = in_policy[M_POLICY_GPU_FREQ_EFFICIENT];
+        m_f_range = m_f_max - m_f_efficient;
+
+        // Per GPU Frequency Selection
+        for (int domain_idx = 0; domain_idx < M_NUM_GPU; ++domain_idx) {
+            // gpu Compute Activity - Primary signal used for frequency recommendation
+            double gpu_compute_activity = m_gpu_compute_activity.at(domain_idx).value;
+            // gpu Utilization - Used to scale activity for short GPU phases
+            double gpu_utilization = m_gpu_utilization.at(domain_idx).value;
+
+            // Default to F_max
+            double f_request = m_f_max;
+
+            if (!std::isnan(gpu_compute_activity)) {
+                // Boundary Checking
+                gpu_compute_activity = std::min(gpu_compute_activity, 1.0);
```

Comment:
Should this be an error condition?  When `gpu_compute_activity` is > 1.0?

Reply:
gpu_compute_activity is provided by DCGM through the nv-hostengine process.  Depending on how nv-hostengine is configured and how noisy the data hw & nv-hostengine return there's the possibility the values could exceed 1.0.  When seen in testing this behavior was transient so the approach of simply clamping to the maximum expected value of 1.0 was used vs GEOPM throwing an error due to bad data inputs.

The end result of gpu_compute_activity > 1.0 is the algorithm selecting the maximum frequency (being performance biased) provided by the user/admin.